Support reading Iceberg split with equality deletes

yingsu00 opened the pull request:
This PR introduces EqualityDeleteFileReader, which is used to read
Iceberg splits with equality delete files.

yingsu00 commented:
@Yuhta This PR replaces https://github.com/facebookincubator/velox/pull/9063 and is ready for review again. As you suggested, we now update the ScanSpec to insert/remove new columns into it instead of cloning the whole ScanSpec. Thanks a lot for reviewing!

czentgr commented:
@yingsu00 can you please rebase again? There is a conflict with other changes.

yingsu00 commented:
@Yuhta Just rebased, appreciate your review again. Thanks!

rui-mo commented:
cc: @liujiayi771

FelixYBW commented:
> cc: @liujiayi771 Would you like to take a review? Thanks.

@liujiayi771 Can you take a look of the PR if possible? should we add something in Gluten side after this PR merged? It's requested by a Gluten customer.

liujiayi771 commented:
@FelixYBW Yes, Gluten needs to make some minor changes to accommodate this PR. I can perform some tests based on this PR.

rui-mo commented on velox/connectors/hive/iceberg/tests/IcebergReadTest.cpp:
```diff
@@ -225,9 +178,202 @@ class HiveIcebergTest : public HiveConnectorTestBase {
+    //    VELOX_CHECK_GE(numDataColumns, equalityDeleteVector.size());
+    //    VELOX_CHECK_GT(equalityDeleteVector.size(), 0);
+    //
+    //    VELOX_CHECK_LE(equalityFieldIds.size(), numDataColumns);
```
Do we need above checks?

rui-mo commented on velox/dwio/common/ScanSpec.h:
```diff
@@ -399,9 +427,12 @@ class ScanSpec {
   bool makeFlat_ = false;
-  std::unique_ptr<common::Filter> filter_;
+  //<<<<<<< HEAD
+  //  std::unique_ptr<common::Filter> filter_;
```
merge conflict?

liujiayi771 commented on velox/connectors/hive/iceberg/FilterUtil.cpp:
```diff
@@ -0,0 +1,122 @@
+      nullAllowed = true;
+    } else {
+      if constexpr (std::is_same_v<U, Timestamp>) {
+        values.emplace_back(simpleValues->valueAt(i).toMillis());
```
@rui-mo It will create a BigIntRange to filter data values. I'm not quite sure why we need to convert the Timestamp type to bigint here; shouldn't we be using TimestampRange instead?

yingsu00 replied:
@rui-mo @liujiayi771 this toValues() function is copied from velox/functions/prestosql/InPredicate.cpp. We could extract it to velox/common but there isn't a good folder yet. Any good ideas? cc @Yuhta

rui-mo replied:
@yingsu00 I see. The InPredicate is also registered by Spark sql and might needs to be fixed through providing configurable behavior. For the one in the 'hive/iceberg/FilterUtil.cpp', I assume a configuration 'kReadTimestampUnit' in the hiveConfig might help adapt to different precision used by Presto and Spark. It's fine for me to leave a TODO here and focus on the primary implementation first. Thanks.

rui-mo replied:
https://github.com/facebookincubator/velox/issues/11755

liujiayi771 replied:
@rui-mo Is the issue here that the precision of a Timestamp in Spark is in microseconds, and converting it to milliseconds for comparison would result in a loss of precision? For example, Timestamp(1, 999000) and Timestamp(1, 998000) would be considered the same?

rui-mo replied:
https://github.com/facebookincubator/velox/issues/11755#issuecomment-2530864206 cc: @liujiayi771

yingsu00 commented on velox/connectors/hive/iceberg/EqualityDeleteFileReader.cpp:
```diff
@@ -0,0 +1,227 @@
+      pool_(connectorQueryCtx->memoryPool()),
+      ioStats_(ioStats),
+      deleteRowReader_(nullptr) {
+  VELOX_CHECK(deleteFile_.content == FileContent::kEqualityDeletes);
```
To use this I added the following in velox/connectors/hive/iceberg/IcebergDeleteFile.h:
```
fmt::underlying_t<FileContent> format_as(FileContent f) {
  return fmt::underlying(f);
}
```

Yuhta commented on velox/connectors/hive/HiveConnectorUtil.cpp:
```diff
@@ -672,6 +672,7 @@ bool applyPartitionFilter(
       VELOX_FAIL(
           "Bad type {} for partition value: {}", type->kind(), partitionValue);
   }
+  return true;
```
Do not need this

Yuhta replied:
Can we remove it?

Yuhta commented on velox/connectors/hive/SplitReader.h:
```diff
@@ -87,6 +90,8 @@ class SplitReader {
 
   void resetSplit();
 
+  std::shared_ptr<const dwio::common::TypeWithId> baseFileSchema();
```
This is only needed inside `IcebergSplitReader`

Yuhta commented on velox/connectors/hive/HiveDataSource.cpp:
```diff
@@ -215,7 +215,10 @@ std::unique_ptr<SplitReader> HiveDataSource::createSplitReader() {
       executor_,
-      scanSpec_);
+      scanSpec_,
+      remainingFilterExprSet_,
```
You are not really using this inside `IcebergSplitReader` (instead you have your own `deleteExprSet_`), let's avoid sharing and passing it.

Yuhta commented on velox/dwio/common/ScanSpec.h:
```diff
@@ -52,19 +52,33 @@ class ScanSpec {
+    filters_.push_back(std::move(filter));
+  }
+
+  void updateFilter(std::unique_ptr<Filter> newFilter) {
```
Let's call it `pushFilter` and `popFilter` to make it more explicit that such filters are temporary

Yuhta commented on velox/dwio/common/ScanSpec.h:
```diff
@@ -198,13 +212,15 @@ class ScanSpec {
-  ScanSpec* getOrCreateChild(const Subfield& subfield);
+  ScanSpec* getOrCreateChild(const Subfield& subfield, bool isTempNode = false);
+
+  void deleteTempNodes();
```
Since all the temp nodes are top level, can we do it less intrusively by keeping a list of temp node/filter names inside `IcebergSplitReader` and remove them after we've done with it?  Then we don't need the extra states of `isTempNode_` and `hasTempFilter_` here.

yingsu00 replied:
@Yuhta Thanks for the suggestion. Yes in our last discussion we thought we could just add/remove top level columns, but when I looked again I found it might be necessary to still mark the nodes with these temp tags, because the Iceberg delete file may contain subfields. For example, this query
```
-- create table t1 (c_row(c1 integer, c2 integer, c3 integer), c_char char);
-- insert some rows

select c_char
from t1 
where c_row.c2 > 2;
```
The base ScanSpec would contain the c_row child, which in turn has 3 children for c1, c2 and c3. In this case, all the three subfields would be null constants, but only c2 node would have filter `c_row.c2 > 2,` while c1 and c3 node don't have any filters. Now, if the Iceberg equality delete file contains the predicate `c_row.c1 = 1 && c_row.c2 IN {2,3}`, then we need to remove these values by adding a filter `c_row.c1<>1` for the c_row.c1 node, and merge the filter `c_row.c2<>2 && c_row.c2<>3` with existing filter `c_row.c2 > 2` for the c_row.c2 node, so it becomes 'c_row.c2 > 3`.  WHen this split finishes, we need to remove the filter `c_row.c1<>1`  from c_row.c1, and restore the filter for c_row.c2 to its original state, but not delete the whole c_row column from the ScanSpec. Therefore we need to tag the nodes in the ScanSpec to check if it's temp node and has temp filters. Do you think this makes sense?

Yuhta commented on velox/expression/Expr.h:
```diff
@@ -722,6 +724,12 @@ class ExprSet {
       core::ExecCtx* execCtx,
       bool enableConstantFolding = true);
 
+  ExprSet(
```
It seems these 2 are no longer needed with delete expr separated from remaining filter expr

Yuhta reviewed (commented):
The overall looks good, one discussion about how the bookkeeping of temporary scan spec node and filters and the rest are just wiring cleanup.

Yuhta commented on velox/dwio/common/ScanSpec.cpp:
```diff
@@ -101,6 +103,19 @@ void ScanSpec::enableFilterInSubTree(bool value) {
+      it = children_.erase(it);
+    } else {
+      if ((*it)->hasTempFilter()) {
+        (*it)->restoreFilter();
```
Need to reset `hasTempFilter_` here

majetideepak commented:
@nmahadevuni can you help resolve the conflicts? Thanks.

czentgr commented:
This PR also has a test failure in `HashJoin.dynamicFilters` that goes away without these changes. It is easily reproducible locally and can be seen failing in the CI pipeline.

```
[ RUN      ] HashJoinTest.dynamicFilters
I0211 11:45:19.812211 1319372 Compression.cpp:629] Initialized zstd compressor with compression level 7
/root/lakehouse/velox/velox/exec/tests/HashJoinTest.cpp:4682: Failure
Expected: (getInputPositions(task, 1)) < (numRowsProbe * numSplits), actual: 3330 vs 3330
Google Test trace:
/root/lakehouse/velox/velox/exec/tests/HashJoinTest.cpp:4669: hasSpill:false
/root/lakehouse/velox/velox/exec/tests/HashJoinTest.cpp:665: Without Spill
/root/lakehouse/velox/velox/exec/tests/utils/QueryAssertions.cpp:1182: Failure
Failed
Expected 1000, got 3330
2330 extra rows, 0 missing rows
10 of extra rows:
        1
        1
        1
        1
        1
        1
        1
        1
        1
        1

0 of missing rows:

Note: DuckDB only supports timestamps of millisecond precision. If this test involves timestamp inputs, please make sure you use the right precision.
DuckDB query: SELECT t.c1 + 1 FROM t, u WHERE t.c0 = u.c0 AND t.c0 < 500
Google Test trace:
/root/lakehouse/velox/velox/exec/tests/HashJoinTest.cpp:665: Without Spill
/root/lakehouse/velox/velox/exec/tests/HashJoinTest.cpp:4805: Failure
Expected: (getInputPositions(task, 1)) < (numRowsProbe * numSplits), actual: 3330 vs 3330
Google Test trace:
/root/lakehouse/velox/velox/exec/tests/HashJoinTest.cpp:4792: hasSpill:false
/root/lakehouse/velox/velox/exec/tests/HashJoinTest.cpp:665: Without Spill
/root/lakehouse/velox/velox/exec/tests/utils/QueryAssertions.cpp:1182: Failure
Failed
Expected 830, got 2450
1620 extra rows, 0 missing rows
10 of extra rows:
        1
        1
        1
        1
        1
        1
        1
        1
        1
        1

0 of missing rows:

Note: DuckDB only supports timestamps of millisecond precision. If this test involves timestamp inputs, please make sure you use the right precision.
DuckDB query: SELECT t.c1 + 1 FROM t, u WHERE t.c0 = u.c0 AND t.c0 < 200
Google Test trace:
/root/lakehouse/velox/velox/exec/tests/HashJoinTest.cpp:665: Without Spill
/root/lakehouse/velox/velox/exec/tests/utils/QueryAssertions.cpp:1182: Failure
Failed
Expected 830, got 2450
1620 extra rows, 0 missing rows
10 of extra rows:
        1
        1
        1
        1
        1
        1
        1
        1
        1
        1

0 of missing rows:

Note: DuckDB only supports timestamps of millisecond precision. If this test involves timestamp inputs, please make sure you use the right precision.
DuckDB query: SELECT t.c1 + 1 FROM t WHERE t.c0 IN (SELECT c0 FROM u) AND t.c0 < 200
Google Test trace:
/root/lakehouse/velox/velox/exec/tests/HashJoinTest.cpp:665: Without Spill
[  FAILED  ] HashJoinTest.dynamicFilters (4442 ms)
[----------] 1 test from HashJoinTest (4443 ms total)
```
```

nmahadevuni commented:
Thanks @czentgr. I have fixed the issue.

yingsu00 commented:
@Yuhta @rui-mo Will you be able to review again? Thanks!

yingsu00 commented:
@Yuhta @rui-mo We've updated this PR, added a separate commit with a fix for the nested subfield filter, and a new test to show the necessity of recursively adding/removing the filter on the nested column in ScanSpec. Would appreciate your review again! Thank you.

jinchengchenghh commented on velox/dwio/common/TypeWithId.cpp:
```diff
@@ -98,6 +112,20 @@ const std::shared_ptr<const TypeWithId>& TypeWithId::childAt(
 
+const std::shared_ptr<const TypeWithId>& TypeWithId::childByFieldId(
+    uint32_t fieldId) const {
+  std::vector<std::string> childNames;
```
childNames and childTypes is not used

jinchengchenghh commented on velox/connectors/hive/iceberg/FilterUtil.cpp:
```diff
@@ -0,0 +1,122 @@
+    vector_size_t size) {
+  auto simpleValues = valuesVector->as<SimpleVector<U>>();
+
+  bool nullAllowed = false;
```
nullAllowed -> hasNulls

rui-mo reviewed (commented):
Thanks!

jinchengchenghh commented:
Please describe the restriction in PR description, e.g. not support hugeint, float, spark neq, timestamp

jinchengchenghh commented:
If the user queries like `select id, data, _deleted from table`, how can we return the metadata column _deleted, so as _pos?

yingsu00 commented on velox/connectors/hive/iceberg/FilterUtil.cpp:
```diff
@@ -0,0 +1,122 @@
+// Read 'size' values from 'valuesVector' starting at 'offset', de-duplicate
+// remove nulls and sort. Return a list of unique non-null values sorted in
+// ascending order and a boolean indicating whether there were any null values.
+template <typename T, typename U = T>
```
@jinchengchenghh This function was copied from velox/functions/prestosql/InPredicate.cpp. In there the typename T is C++ type and U is Velox Type. To reduce code duplication and make the code clearer, I just moved it to velox/type/Filter.h, and renamed it as `deDuplicateValues` in a separate commit "refactor: Move toValues from InPredicate.cpp to Filter.h"

yingsu00 commented on velox/connectors/hive/iceberg/FilterUtil.cpp:
```diff
@@ -0,0 +1,122 @@
+  values.reserve(size);
+
+  for (auto i = offset; i < offset + size; i++) {
+    if (simpleValues->isNullAt(i)) {
```
> Please consider mayHaveNulls

mayHaveNulls() may not be exactly accurate. Here we are building the filter functions and have to be 100% accurate. But I agree that we should be able to improve this loop when mayHaveNulls() is false. But since this is a pre-existing function, I think it's better to do it in later PRs. I created https://github.com/facebookincubator/velox/issues/12659 for it.

yingsu00 commented on velox/connectors/hive/iceberg/IcebergDeleteFile.cpp:
```diff
@@ -0,0 +1,40 @@
+
+namespace facebook::velox::connector::hive::iceberg {
+
+fmt::underlying_t<FileContent> format_as(FileContent f) {
```
> format_as -> formatAs

This is the formatter for fmt library and has to be named as such. But I will add a comment there.
```
/// Formatter for fmt.
fmt::underlying_t<FileContent> format_as(FileContent f);
```

yingsu00 commented on velox/connectors/hive/iceberg/IcebergSplitReader.cpp:
```diff
@@ -63,20 +72,81 @@ void IcebergSplitReader::prepareSplit(
+  }
+
+  if (!subfieldFilters.empty()) {
+    for (auto iter = subfieldFilters.begin(); iter != subfieldFilters.end();
```
> Please use `for (auto it : mp) { `

Ok. changed to `for (const auto& [key, filter] : subfieldFilters) {`

yingsu00 commented on velox/dwio/common/TypeWithId.cpp:
```diff
@@ -74,15 +74,29 @@ int countNodes(const TypePtr& type) {
+  };
+
   for (int i = 0, size = type->size(); i < size; ++i) {
+    // There is no guarantee that the spec contains
```
> incomplete

Changed to `// Recursively create children if the extra field is required.`

yingsu00 commented on velox/connectors/hive/iceberg/EqualityDeleteFileReader.cpp:
```diff
@@ -0,0 +1,222 @@
+            std::make_shared<FieldAccessTypedExpr>(type, name));
+        isNotEqualInputs.push_back(std::make_shared<ConstantTypedExpr>(value));
+        auto isNotEqualExpr =
+            std::make_shared<CallTypedExpr>(BOOLEAN(), isNotEqualInputs, "neq");
```
> Spark does not register the `neq` function but contains `not` and `equalto`

Good catch! I don't have an ideal way to solve this now. How do you want to address it? For now I added the following TODO comment in the code:
```
// TODO: generalize this to support different engines. Currently, only
        // Presto "neq" is supported. Spark does not register the "neq" function
        // but does support "not" and "equalto" functions.
```

yingsu00 replied:
Created https://github.com/facebookincubator/velox/issues/12667

yingsu00 commented on velox/connectors/hive/iceberg/FilterUtil.cpp:
```diff
@@ -0,0 +1,122 @@
+      filter = createNegatedBigintValuesFilter<int32_t>(elements, offset, size);
+      break;
+    case TypeKind::SMALLINT:
+      filter = createNegatedBigintValuesFilter<int16_t>(elements, offset, size);
```
> We convert all the data to int64_t, here might be a performance issue

Note that Velox Filters only has 64bit or 128 bit implementations for all integral types. Maybe in the future we can improve this but we have to live with it for now. Btw. this code was also copied from `InPredicate::create()`. 
```
switch (elementType->kind()) {
      case TypeKind::HUGEINT:
        filter = createHugeintValuesFilter<int128_t>(elements, offset, size);
        break;
      case TypeKind::BIGINT:
        filter = createBigintValuesFilter<int64_t>(elements, offset, size);
        break;
      case TypeKind::INTEGER:
        filter = createBigintValuesFilter<int32_t>(elements, offset, size);
        break;
      case TypeKind::SMALLINT:
        filter = createBigintValuesFilter<int16_t>(elements, offset, size);
        break;
      case TypeKind::TINYINT:
        filter = createBigintValuesFilter<int8_t>(elements, offset, size);
        break;
```

yingsu00 commented:
> Please describe the restriction in PR description, e.g. not support hugeint, float, spark neq, timestamp



> If the user queries like `select id, data, _deleted from table`, how can we return the metadata column _deleted, so as _pos?

Currently Presto Java and C++ only supports $path and $data_sequence_number metadata columns. See https://prestodb.io/docs/current/connector/iceberg.html#extra-hidden-metadata-columns. We will be adding more in the future.

yingsu00 commented:
@rui-mo @Yuhta The PR is updated, could you please review again? Thank you!

rui-mo commented on velox/connectors/hive/iceberg/EqualityDeleteFileReader.h:
```diff
@@ -0,0 +1,73 @@
+using SubfieldFilters =
+    std::unordered_map<common::Subfield, std::unique_ptr<common::Filter>>;
+
+class EqualityDeleteFileReader {
```
It would be nice to document this class and its member variables.

yingsu00 replied:
> It would be nice to document this class and its member variables.

@rui-mo Do you mean adding comments in the code?

rui-mo replied:
Yes, thanks.

rui-mo reviewed (commented):
Thanks! Looks good % some nits. cc @Yuhta for further review.

Yuhta commented:
@yingsu00 Do we still want to proceed with this change or we just start a new IceBergDataSource without directly inheriting the abstract `DataSource`?

yingsu00 commented:
> @yingsu00 Do we still want to proceed with this change or we just start a new `IcebergDataSource` directly inheriting the abstract `DataSource` without going through `HiveDataSource`?

@Yuhta Hi Jimmy thanks for reviewing again! We are compiling the new IcebergConnector PR and it will be out in a couple of weeks. I think we can wait for that one is merged first, then we will re-submit this PR to be in the new IcebergConnector. Thank you in advance for reviewing the upcoming PRs! Your review is very much appreciated.
